### PR TITLE
tslint to eslint migration

### PR DESCRIPTION
**Pull request by jrieken**
This PR explores to migrate our [custom tslint rules](https://github.com/microsoft/vscode/blob/da25e0802a1057404b98dfcb564a460a160144d2/build/lib/tslint) to eslint rules and adopts eslint in general.

Rules that we have today and that need migration

- [x] duplicateImportsRule.ts (*comes with eslint*)
- [x] importPatternsRule.ts (*this PR*)
- [x] layeringRule.ts (9819da68aba0d5a40489aacbcccad7d3a8d1a390)
- ~~noDomGlobalsRule.ts~~ (*moved to custom solution*)
- [x] noNewBufferRule.ts ([eslint rule](https://github.com/eslint/eslint/blob/master/lib/rules/no-buffer-constructor.js), 291ee0058a8f4dff8e42bb0c5b3a42a2a1f54c43)
- [x] noNlsInStandaloneEditorRule.ts (2ea9132cab688d99dd2c9fa57fb12f3b91b1d689)
- ~~noNodejsGlobalsRule.ts~~ (*moved to custom solution*)
- [x] noStandaloneEditorRule.ts (b05b48126571d8121dbbf55c6dde6e966dbf762c)
- [x] noUnexternalizedStringsRule.ts (a61f38eee52329c90d76ce817c527c99012fdb08)
- [x] translationRemindRule.ts (a3bd6046273957d2b0ebe3056abf08de45d7f036)

Other work packages that are required to migrate away from eslint

- ~~investigate using https://eslint.org/docs/rules/no-restricted-imports~~ (*our needs are too special*)
- [x] find estree-equivalent of  `import foo = require('...')`
- [x] use ruledir-setting instead of plugin-approach?
- [x] migrate our existing tslint-config to eslint (*using tslint-to-eslint-config result plus tweaks)*
- [x] integrate into build 
- [x] integrate into precommit hook
- [x] editor integration using our extension

### CODE REVIEW

**Comment by Stanzilla** [comment]
Just curious, what's the reason behind the move?

**Comment by mattacosta** [comment]
@Stanzilla The TSLint project is winding down and will no longer be maintained in the future. See https://github.com/palantir/tslint/issues/4534.

**Comment by bpasero** [comment]
Pushed https://github.com/microsoft/vscode/commit/7909462634a44f09e7f805bdeb9c082fd533e7e5 to move the globals rules (node.js, DOM) into a custom solution that uses the TS API directly. I could not get that to work with a `tsconfig.json`. 

Btw there is a eslint rule for "new Buffer": https://github.com/eslint/eslint/blob/master/lib/rules/no-buffer-constructor.js

**Comment by jrieken** [comment]
@bpasero Won't this rule help you? https://eslint.org/docs/rules/no-restricted-globals

**Comment by jrieken** [comment]
current rule mapping/migration state

`class-name` <=> `@typescript-eslint/class-name-casing`
`curly` <=> `curly`
`duplicate-imports` <=> `no-duplicate-imports`
`import-patterns` <=> `code-import-patterns`
`label-position` <=> `no-unused-labels (similar)`
`layering` <=> `code-layering`
`no-arg` <=> `no-caller`
`no-construct` <=> `no-new-wrappers`
`no-debugger` <=> `no-debugger`
`no-duplicate-super` <=> `constructor-super`
`no-duplicate-switch-case` <=> `no-duplicate-case`
`no-duplicate-variable` <=> `no-redeclare (OFF, unexpected warnings)`
`no-eval` <=> `no-eval`
`no-for-in-array` <=> `@typescript-eslint/no-for-in-array (OFF, need type info)`
`no-new-buffer` <=> `no-buffer-constructor`
`no-nls-in-standalone-editor` <=> `code-no-nls-in-standalone-editor`
`no-redundant-jsdoc` <=> `jsdoc/no-types (OFF, unexpected warnings)`
`no-restricted-globals` <=> `N/A`
`no-sparse-arrays` <=> `no-sparse-arrays`
`no-standalone-editor` <=> `code-no-standalone-editor`
`no-string-throw` <=> `no-throw-literal`
`no-unexternalized-strings` <=> `code-no-unexternalized-strings`
`no-unsafe-finally` <=> `no-unsafe-finally`
`no-unused-expression` <=> `no-unused-expressions (OFF, unexpected warnings)`
`no-var-keyword` <=> `no-var`
`number-literal-format` <=> `N/A`
`semicolon` <=> `@typescript-eslint/semi`
`translation-remind` <=> `code-translation-remind`
`triple-equals` <=> `eqeqeq`

File: src/vs/platform/log/common/fileLogService.ts [hunk 1]
```diff
@@ -173,7 +173,7 @@ export class FileLoggerService extends Disposable implements ILoggerService {
 	getLogger(resource: URI): ILogger {
 		let logger = this.loggers.get(resource.toString());
 		if (!logger) {
-			logger = new BufferLogService, this.logService.getLevel();
+			logger = new BufferLogService(this.logService.getLevel());
```
Comment (jrieken) [inline src/vs/platform/log/common/fileLogService.ts]: fyi @sandy081

File: src/vs/workbench/contrib/files/browser/editors/textFileSaveErrorHandler.ts [hunk 2]
```diff
@@ -180,7 +180,7 @@ export class TextFileSaveErrorHandler extends Disposable implements ISaveErrorHa
 		// Show message and keep function to hide in case the file gets saved/reverted
 		const actions: INotificationActions = { primary: primaryActions, secondary: secondaryActions };
 		const handle = this.notificationService.notify({ severity: Severity.Error, message, actions });
-		Event.once(handle.onDidClose)(() => { dispose(primaryActions), dispose(secondaryActions); });
+		Event.once(handle.onDidClose)(() => { dispose(primaryActions); dispose(secondaryActions); });
```
Comment (jrieken) [inline src/vs/workbench/contrib/files/browser/editors/textFileSaveErrorHandler.ts]: fyi @mjbvz

**Comment by bpasero** [comment]
I think `no-restricted-globals` is still valid to have (and not covered by my solution), since it [exists in ESLint](https://eslint.org/docs/rules/no-restricted-globals), I can bring it back.

File: src/vs/platform/log/common/fileLogService.ts (hunk 1 above)
Comment (sandy081) [reply src/vs/platform/log/common/fileLogService.ts]: Good catch by the linter

**Pull request merged**